### PR TITLE
Sb BinaryData - Binary representation of data from different sources

### PR DESCRIPTION
**Pull request by hemanttanwar**
- Binary representation of data from different sources
- Support byte[], InputStream , ByteBuffer, Object
- Support ObjectSerializer to serialize/deserialize Object

Java: API View : https://apiview.dev/Assemblies/Review/22274ac987a04a7589b6d5e6ee6b7554#com.azure.core.experimental.util.BinaryData

.Net API View : https://apiview.dev/Assemblies/Review/8f2e023e377f4627a4ec815f6e11d013#Azure.Core.BinaryData

### CODE REVIEW

File: sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/util/BinaryData.java [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+        // This exists, so no one is able to create instance, user need to use static function to create instances.
+        this.data = null;
+    }
+
+    /**
+     * Create instance of {@link BinaryData} given the data.
+     * @param data to represent as bytes.
+     */
+    BinaryData(byte[] data) {
+        this.data = data;
```
Comment (alzimmermsft) [inline sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/util/BinaryData.java:59]: Should we clone this? In other APIs we usually make a copy of the `byte[]`.

File: sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/util/BinaryData.java [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+     * @throws UncheckedIOException If any error in reading from {@link InputStream}.
+     * @throws NullPointerException if {@code inputStream} is null.
+     * @return {@link BinaryData} representing the binary data.
+     */
+    public static BinaryData fromStream(InputStream inputStream) {
+        Objects.requireNonNull(inputStream, "'inputStream' cannot be null.");
+
+        final int bufferSize = 1024;
+        try {
+            ByteArrayOutputStream dataOutputBuffer = new ByteArrayOutputStream();
```
Comment (alzimmermsft) [inline sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/util/BinaryData.java:100]: Once this moves into `azure-core` we have a specialized `ByteArrayOutputStream` type that allows access to its internal buffer without copying when `toByteArray` is called.

File: sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/util/BinaryData.java [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+            int nRead;
+            byte[] data = new byte[bufferSize];
+            while ((nRead = inputStream.read(data, 0, data.length)) != -1) {
+                dataOutputBuffer.write(data, 0, nRead);
+            }
+
+            return fromBytes(dataOutputBuffer.toByteArray());
+        } catch (IOException ex) {
+            ClientLogger logger = new ClientLogger(BinaryData.class);
```
Comment (alzimmermsft) [inline sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/util/BinaryData.java:109]: Given this class only has static utility methods it should be fine to use a static logger and suppress the Checkstyle rule.

File: sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/util/BinaryData.java (hunk 2 above)
Comment (hemanttanwar) [reply sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/util/BinaryData.java:100]: yes we can do this when we move it azure-core

**Review by alzimmermsft — APPROVED** [review]

**Review by srnagar — APPROVED** [review]

**Pull request merged**